usb/device/USBCDC_ECM/USBCDC_ECM.cpp

```diff
@@ -0,0 +1,496 @@
+    USBDevice::write_start(_int_in, request, sizeof(request));
+    uint32_t flags = _flags.wait_any(FLAG_INT_DONE | FLAG_DISCONNECT, osWaitForever, false);
+    if (flags & FLAG_DISCONNECT) {
+        ret = false;
+    }
+    USBDevice::write_finish(_int_in);
+
+    _write_mutex.unlock();
+    return ret;
+}
+
+bool USBCDC_ECM::_notify_connection_speed_change(uint32_t up, uint32_t down)
+{
+    _write_mutex.lock();
+
+    bool ret = true;
+    struct notification_t {
+        uint8_t request[8];
+        uint32_t up;
+        uint32_t down;
+    };
+
+    notification_t notification;
+    memset(&notification, 0, sizeof(notification));
+
+    notification.request[0] = 0xA1;
+    notification.request[1] = CONNECTION_SPEED_CHANGE;
+    notification.request[2] = 0x00;
+    notification.request[3] = 0x00;
+    notification.request[6] = 0x08;
+    notification.up = up;
+    notification.down = down;
+
+    _flags.clear(FLAG_INT_DONE);
+    USBDevice::write_start(_int_in, (uint8_t *)&notification, sizeof(notification));
+    uint32_t flags = _flags.wait_any(FLAG_INT_DONE | FLAG_DISCONNECT, osWaitForever, false);
+    if (flags & FLAG_DISCONNECT) {
+        ret = false;
+    }
+    USBDevice::write_finish(_int_in);
+
+    _write_mutex.unlock();
+    return ret;
+}
+
+bool USBCDC_ECM::_write_bulk(uint8_t *buffer, uint32_t size)
+{
+    bool ret = true;
+
+    _flags.clear(FLAG_WRITE_DONE);
+    USBDevice::write_start(_bulk_in, buffer, size);
+    uint32_t flags = _flags.wait_any(FLAG_WRITE_DONE | FLAG_DISCONNECT, osWaitForever, false);
+    if (flags & FLAG_DISCONNECT) {
+        ret = false;
+    }
+
+    USBDevice::write_finish(_bulk_in);
+    return ret;
+}
+
+bool USBCDC_ECM::send(uint8_t *buffer, uint32_t size)
+{
+    _write_mutex.lock();
+    bool ret = true;
+    uint32_t sent = 0;
+    uint32_t data_size = 0;
+
+    if (size > MAX_SEGMENT_SIZE) {
+        _write_mutex.unlock();
+        mbed_error_printf("Buffer size is too large\n");
+        return false;
+    }
+
+    uint32_t max_packet = USBDevice::endpoint_max_packet_size(_bulk_in);
+
+    while (size - sent > 0) {
+        data_size = (size - sent > max_packet) ? max_packet : size - sent;
+        if (_write_bulk(buffer + sent, data_size)) {
+            sent += data_size;
+        } else {
+            _write_mutex.unlock();
+            return false;
+        }
+    }
+
+    /* Send zero length packet */
+    if (size % max_packet == 0) {
+        uint8_t buf = 0;
+        ret = _write_bulk(&buf, 0);
+    }
+
+    _write_mutex.unlock();
+    return ret;
+}
+
+void USBCDC_ECM::callback_request(const setup_packet_t *setup)
+{
+    assert_locked();
+
+    RequestResult result = PassThrough;
+    uint8_t *data = NULL;
+    uint32_t size = 0;
+
+    if (setup->bmRequestType.Type == CLASS_TYPE) {
+        //printf("In USBCallback_request: CLASS specific Request: %02x\n", setup->bRequest);
+        switch (setup->bRequest) {
+            case SET_ETHERNET_MULTICAST_FILTERS:
```

Comment:
Should these be forwarded to the user of this call as a callback? Alternatively, if this class took in a an `EMAC` phy then you could call the appropriate functions. @kjbracey-arm what are your thoughts on this?